crates/re_types_builder/src/codegen/cpp/mod.rs

```diff
@@ -1090,69 +1079,90 @@ fn single_field_constructor_methods(
-        // provide a constructor for that single field as well.
-        //
-        // Note that we previously we tried to do a general forwarding constructor via variadic templates,
-        // but ran into some issues when init archetypes with initializer lists.
-        if let Type::Object(field_type_fqname) = &obj_field.typ {
-            let field_type_obj = &objects[field_type_fqname];
-            if field_type_obj.fields.len() == 1 {
-                let inner_field = &field_type_obj.fields[0];
-                let arg_name = format_ident!("arg");
-
-                if let Type::Array { elem_type, length } = &inner_field.typ {
-                    // Reminder: Arrays can't be passed by value, they decay to pointers. So we pass by reference!
-                    // (we could pass by pointer, but if an extension wants to add smaller array constructor these would be ambiguous then!)
-                    let length_quoted = quote_integer(length);
-                    let element_type = quote_element_type(hpp_includes, elem_type);
-                    methods.push(Method {
-                        declaration: MethodDeclaration::constructor(quote! {
-                            #type_ident(const #element_type (&#arg_name)[#length_quoted]) : #field_ident(#arg_name)
-                        }),
-                        ..Method::default()
-                    });
-                } else {
-                    let argument = quote_variable(hpp_includes, inner_field, &arg_name);
-                    methods.push(Method {
-                        declaration: MethodDeclaration::constructor(quote! {
-                            #type_ident(#argument) : #field_ident(std::move(#arg_name))
-                        }),
-                        ..Method::default()
-                    });
-                }
+    fn parameter_decl(
+        hpp_includes: &mut Includes,
+        obj_field: &ObjectField,
+        objects: &Objects,
+        param_ident: &Ident,
+    ) -> TokenStream {
+        let typ = quote_field_type(hpp_includes, obj_field);
+        if obj_field.typ.has_default_destructor(objects) {
+            // Could always do `const&` passing since in this path we'll do a copy, but it's usually frowned upon for super simple types like `int`
+            if obj_field.typ.is_plural() || obj_field.typ.fqname().is_some() {
+                quote! { const #typ& #param_ident }
+            } else {
+                quote! { #typ #param_ident }
             }
+        } else {
+            // For anything that's worth moving we pass by value. Rationale:
+            // - If it was a temporary it gets moved into the value and then moved again into the field.
+            // - If it was a lvalue it gets copied into the value and then moved into the field.
+            //
+            // If the type is "primitive" however, we essentially copy anyways, so we pass by const-ref.
+            quote! { #typ #param_ident }
```

Comment:
Please expand on the differences between the `has_default_destructor` branches - why would that matter? Why not always pass by value and use `std::move` in the initializer list?